corehq/apps/enterprise/iterators.py

```diff
@@ -1,7 +1,197 @@
+            'app_name': self.app_lookup.resolve_app_id_to_name(domain, form['app_id']) or _('App not found'),
+            'username': form['form']['meta']['username'],
+            'domain': domain
+        }
+
+
+class AppIdToNameResolver:
+    def __init__(self):
+        self.domain_lookup_tables = {}
+
+    def resolve_app_id_to_name(self, domain, app_id):
+        if 'domain' not in self.domain_lookup_tables:
+            domain_apps = get_brief_apps_in_domain(domain)
+            self.domain_lookup_tables[domain] = {a.id: a.name for a in domain_apps}
+
+        return self.domain_lookup_tables[domain].get(app_id, None)
+
+
+def multi_domain_form_generator(
+        domains, start_date, end_date, last_domain=None, last_time=None, last_id=None, limit=None):
+    domain_index = domains.index(last_domain) if last_domain else 0
+
+    remaining = limit
+
+    def _get_domain_iterator(last_time=None, last_id=None):
+        if domain_index >= len(domains):
+            return None
+        domain = domains[domain_index]
+        return domain_form_generator(domain, start_date, end_date, last_time, last_id, limit=remaining)
+
+    current_iterator = _get_domain_iterator(last_time, last_id)
+
+    while current_iterator:
+        for form in current_iterator:
+            yield form
+            if remaining:
+                remaining -= 1
+                if remaining == 0:
+                    return
+        domain_index += 1
+        if domain_index >= len(domains):
+            return
+        current_iterator = _get_domain_iterator()
+
+
+def domain_form_generator(domain, start_date, end_date, last_time=None, last_id=None, limit=None):
+    if not last_time:
+        last_time = datetime.now()
+
+    remaining = limit
+
+    while True:
+        query = create_domain_forms_query(domain, start_date, end_date, last_time, last_id, limit=remaining)
+        results = query.run()
+        for form in results.hits:
+            last_form_fetched = form
+            yield last_form_fetched
+
+        num_fetched = len(results.hits)
+
+        if num_fetched >= results.total or (remaining and num_fetched >= remaining):
+            break
+        else:
+            if remaining:
+                remaining -= num_fetched
+                assert remaining > 0
+            last_time = last_form_fetched['received_on']
+            last_id = last_form_fetched['_id']
+
+
+def create_domain_forms_query(domain, start_date, end_date, last_time, last_id, limit=None):
+    query = (
+        FormES()
+        .domain(domain)
+        .user_type('mobile')
+        .submitted(gte=start_date, lte=end_date)
+    )
+
+    if limit:
+        query = query.size(limit)
+
+    query.es_query['sort'] = [
+        {'inserted_at': {'order': 'desc'}},
+        {'doc_id': 'asc'}
+    ]
+
+    if last_id:
+        query = query.filter(filters.OR(
+            filters.AND(
+                filters.term('inserted_at', es_format_datetime(last_time)),
```

Comment:
can you assign a variable name to this to make this logic easier to understand at a glance? and/or add comments?